### PR TITLE
Database

### PR DESCRIPTION
**Pull request by FranziGraml**
In this PR I connected my database (MongoDB Atlas) with Add and Delete Button.
Sorry for all the commits at the end but my github checks kept failing! Sorry!!!

### CODE REVIEW

File: pages/api/post/[id].js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import Post from '../../../src/models/Post';
+
+export default async function handler(req, res) {
+	const { id } = req.query;
+
+	if (req.method === 'DELETE') {
+		const deletedPost = await Post.findByIdAndDelete(id);
+		res.status(200).json({ message: 'post deleted', post: deletedPost });
```
Comment (F-Kirchhoff) [inline pages/api/post/[id].js:8]: what happens if the deletion fails? I guess you can add a try and catch block here.

File: pages/api/post/[id].js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+	} else {
+		const singleCard = await Post.findById(id);
```
Comment (F-Kirchhoff) [inline pages/api/post/[id].js:10]: Even when a 'POST' or 'PUT' request is made this code is executed. Please add another else if block to check whether the method is 'GET'. (You could use a `switch` statement as well).

File: pages/api/post/create.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+			name: data.name,
+			content: data.post,
+			mail: data.mail,
+			mobile: data.mobile,
+			postDate: data.postDate,
+		});
+
+		response.status(200).json({ message: 'post created', post: newPost });
+	} else {
+		response.status(400).json({ error: 'wrong method' });
```
Comment (F-Kirchhoff) [inline pages/api/post/create.js:25]: the correct status code is 405.

File: pages/api/post/create.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+		const newPost = await Post.create({
+			name: data.name,
+			content: data.post,
+			mail: data.mail,
+			mobile: data.mobile,
+			postDate: data.postDate,
+		});
+
+		response.status(200).json({ message: 'post created', post: newPost });
```
Comment (F-Kirchhoff) [inline pages/api/post/create.js:23]: Here you could also add error handling if you like.

File: pages/posts.js [hunk 5]
```diff
@@ -1,41 +1,41 @@
...
+import Icon from '../src/Components/UI/Icons/icons';
+import { SWRConfig } from 'swr';
+import { swrFetcher } from '../src/Components/lib/swr-Fetcher';
+import RenderPosts from '../src/Components/Form/RenderPosts';
 
-export function getStaticProps() {
-	const data = getPosts();
+export async function getStaticProps() {
+	const posts = await getPosts();
+	console.log('posts: ' + JSON.stringify(posts));
```
Comment (F-Kirchhoff) [inline pages/posts.js:14]: Please remove all logs before merging 🕵️

File: pages/posts.js [hunk 6]
```diff
@@ -1,41 +1,41 @@
...
-				<Form posts={posts} onSetPosts={setPosts} onSetIsFormActive={setIsFormActive} />
-			</FormWrapper>
-			<RenderPosts posts={posts} onSetPosts={setPosts} />
+			<SWRConfig value={{ fetcher: swrFetcher, fallback }}>
+				<ButtonAdd type="button" onClick={() => setIsFormActive(prevCheck => !prevCheck)}>
+					<Icon variant="plus" />
+				</ButtonAdd>
+				<FormWrapper isFormActive={isFormActive}>
+					<Form onSetIsFormActive={setIsFormActive} />
+				</FormWrapper>
```
Comment (F-Kirchhoff) [inline pages/posts.js:36]: Normally we want to keep the logic for setting states in the page file and not let the component do the work. I guess its better when you create a little function that automatically sets the state value here in the page component and pass this function down to the Form.

File: src/Components/Form/DeleteButton.js [hunk 7]
```diff
@@ -1,13 +1,24 @@
...
 
 	return (
-		<ButtonDelete type="button" onClick={() => deletePost()}>
+		<ButtonDelete
+			type="button"
+			onClick={async () => {
+				const response = await fetch('/api/post/' + id, {
+					method: 'DELETE',
+				});
+				console.log(await response.json());
```
Comment (F-Kirchhoff) [inline src/Components/Form/DeleteButton.js:17]: 🕵️

File: src/Components/Form/Form.js [hunk 8]
```diff
@@ -3,39 +3,43 @@ import FormStyled from '../UI/Form/Form.styles';
...
 import ButtonSubmit from '../UI/Form/Button/Submitbutton.styles';
 import ErrorBox from '../UI/Form/ErrorBox.styles';
+import { useRouter } from 'next/router';
 
-export default function Form({ posts, onSetPosts, onSetIsFormActive }) {
+export default function Form({ onSetIsFormActive }) {
 	const [nameValue, setNameValue] = useState('');
 	const [postValue, setPostValue] = useState('');
 	const [mailValue, setMailValue] = useState('');
 	const [mobileValue, setMobileValue] = useState('');
```
Comment (F-Kirchhoff) [inline src/Components/Form/Form.js:17]: Maybe you can refactor this into one object state at some point?

File: src/Components/Form/RenderPosts.js [hunk 9]
```diff
@@ -1,23 +1,31 @@
...
-							posts={posts}
-							onSetPosts={onSetPosts}
-						/>
-					</ul>
-				);
-			})}
+			{posts
+				.sort((a, b) => b.postDate - a.postDate)
+				.map(post => {
+					return (
```
Comment (F-Kirchhoff) [inline src/Components/Form/RenderPosts.js:16]: you could use an implicit return here: 
```
post => (
  <ul key = ...>
  ...
  </ul>
)
```

File: src/Services/initial-post-db.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    "mail": "hallo.world@com",
+    "mobile": "25598078"
+  },
+  {
+    "name": "Muc-web",
+    "content": "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. ",
+    "mail": "hallo.terra@com",
+    "mobile": "648893678"
+  }
+]
```
Comment (F-Kirchhoff) [inline src/Services/initial-post-db.json:20]: You don't need this any longer if you have a db right?

**Review by F-Kirchhoff — CHANGES_REQUESTED** [review]
Hey Franzi, looking good so far. I have only some minor suggestions about error handling and state management. Besides that its top!

File: src/Components/Form/Form.js [hunk 11]
```diff
@@ -3,39 +3,43 @@ import FormStyled from '../UI/Form/Form.styles';
...
+					postDate: post_date,
+				}),
+			});
 			setNameValue('');
 			setPostValue('');
 			setMailValue('');
 			setMobileValue('');
 			onSetIsFormActive(false);
 			setIsError(false);
+			console.log(await response.json);
```
[the remark this answers is not in the document]
Comment (FranziGraml) [reply src/Components/Form/Form.js:41]: same here

**Comment by FranziGraml** [comment]
Thanks a lot for the nice and helpful feedback!

File: src/Components/Form/RenderPosts.js [hunk 12]
```diff
@@ -1,23 +1,31 @@
...
+						<ul key={nanoid()}>
```
Comment (doemser) [inline src/Components/Form/RenderPosts.js:17]: take your `post.id` instead

File: src/Components/Form/Form.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (doemser) [reply src/Components/Form/Form.js:41]: You did not remove that?

File: pages/posts.js (hunk 6 above)
Comment (doemser) [reply pages/posts.js:36]: Here you pass down the `setIsFormActive` function directly into your `Form`. This is something you should avoid.

Instead do something like this:

```
function deactivateForm() {
setIsFormActive(!isFormActive);
}
```

An then pass it down like this:

`<Form onSetIsFormActive={deactivateForm} />`

This way you don't separate the logic from the useState, means: You can see which changes to the useState can be applied without leaving the place where the useState was written.

This procedure is very important and is called "Lifting state up". Here is a [useful Link](https://reactwithhooks.netlify.app/docs/lifting-state-up.html)

File: pages/posts.js (hunk 5 above)
Comment (doemser) [reply pages/posts.js:14]: Still there? ;)

File: pages/api/post/create.js (hunk 3 above)
Comment (doemser) [reply pages/api/post/create.js:25]: What Felix said

File: pages/api/post/create.js (hunk 4 above)
Comment (doemser) [reply pages/api/post/create.js:23]: Error-handling still missing :P

**Review by doemser — CHANGES_REQUESTED** [review]
You replied with "thumbs up" to most of the changes Felix requested, but did not implement them.

**Comment by FranziGraml** [comment]
Hi Dominik
Thanks for you reviewing. I hope that this time my changes were loaded correctly.

File: pages/posts.js [hunk 13]
```diff
@@ -25,14 +23,18 @@ export async function getStaticProps() {
...
+
 	return (
 		<main>
 			<SWRConfig value={{ fetcher: swrFetcher, fallback }}>
 				<ButtonAdd type="button" onClick={() => setIsFormActive(prevCheck => !prevCheck)}>
 					<Icon variant="plus" />
 				</ButtonAdd>
 				<FormWrapper isFormActive={isFormActive}>
-					<Form onSetIsFormActive={setIsFormActive} />
+					<Form onSetFormActiveFalse={() => setFormActiveFalse()} />
```
Comment (F-Kirchhoff) [inline pages/posts.js:37]:
```suggestion
					<Form onSetFormActiveFalse={setFormActiveFalse} />
```
This works as well and is a bit shorter. But your solution is nice too.

File: src/Components/Form/DeleteButton.js [hunk 14]
```diff
@@ -11,10 +11,9 @@ export default function DeleteButton({ id }) {
 		<ButtonDelete
 			type="button"
 			onClick={async () => {
-				const response = await fetch('/api/post/' + id, {
+				const _response = await fetch('/api/post/' + id, {
```
Comment (F-Kirchhoff) [inline src/Components/Form/DeleteButton.js:14]: if you don't want to use the response you can just omit the constant declaration:
```suggestion
				await fetch('/api/post/' + id, {
```

File: src/Components/Form/RenderPosts.js [hunk 15]
```diff
@@ -1,6 +1,6 @@
 import PostCard from '../PostCard/PostCard';
 import useSWR from 'swr';
-import { nanoid } from 'nanoid';
+//import { nanoid } from 'nanoid';
```
Comment (F-Kirchhoff) [inline src/Components/Form/RenderPosts.js:3]: Please remove code instead of commenting it out ;)

File: src/Components/Form/DeleteButton.js (hunk 14 above)
Comment (FranziGraml) [reply src/Components/Form/DeleteButton.js:14]: Good to know. I would still like to keep it as I was very pleased with my solution.

**Pull request merged**